nisapi/clean.py

```diff
@@ -0,0 +1,253 @@
+import polars as pl
+import polars.testing
+
+"""Data schema to be used for all datasets"""
+data_schema = pl.Schema(
+    [
+        ("vaccine", pl.String),
+        ("geographic_type", pl.String),
+        ("geographic_value", pl.String),
+        ("demographic_type", pl.String),
+        ("demographic_value", pl.String),
+        ("indicator_type", pl.String),
+        ("indicator_value", pl.String),
+        ("week_ending", pl.Date),
+        ("estimate", pl.Float64),
+        ("ci_half_width_95pct", pl.Float64),
+    ]
+)
+
+
+def clean_dataset(id: str, df: pl.DataFrame) -> pl.DataFrame:
+    """Clean a raw dataset, applying dataset-specific cleaning rules
+
+    Args:
+        id (str): dataset ID
+        df (pl.DataFrame): raw dataset
+
+    Returns:
+        pl.DataFrame: clean dataset
+    """
+    clean = df
+
+    if id not in ["sw5n-wg2p", "ksfb-ug5d"]:
+        raise RuntimeError(f"No cleaning set up for dataset {id}")
+
+    # Drop rows with suppression flags
+    clean = clean.filter(pl.col("suppression_flag") == pl.lit("0"))
+
+    if id == "sw5n-wg2p":
+        # this particular dataset has a bad column name
+        clean = clean.rename({"estimates": "estimate"})
+
+    if id in ["sw5n-wg2p", "ksfb-ug5d"]:
+        clean = (
+            clean.pipe(rename_indicator_columns)
+            # drop unneeded columns
+            .select(data_schema.names())
+            # change most string columns to all lowercase
+            # (but don't change, e.g., the names of states)
+            .with_columns(
+                pl.col(
+                    [
+                        "vaccine",
+                        "geographic_type",
+                        "demographic_type",
+                        "indicator_value",
+                        "indicator_type",
+                    ]
+                ).str.to_lowercase()
+            )
+            # cast types
+            .with_columns(
+                pl.col("week_ending").str.strptime(pl.Datetime, "%Y-%m-%dT%H:%M:%S%.f"),
+                pl.col(["estimate", "ci_half_width_95pct"]).cast(pl.Float64),
+            )
+            # change percents to proportions
+            .with_columns(pl.col(["estimate", "ci_half_width_95pct"]) / 100.0)
+        )
+
+        # check that the date doesn't have any trailing seconds
+        assert (clean["week_ending"].dt.truncate("1d") == clean["week_ending"]).all()
+        clean = clean.with_columns(pl.col("week_ending").dt.date())
+
+        clean = (
+            clean
+            # Change from "national" to "nation", so that types are nouns rather
+            # than adjectives. (Otherwise we would need to change "region" to "regional")
+            .with_columns(
+                pl.col("geographic_type").replace({"national": "nation"}),
+                pl.col("geographic_value").replace({"National": "nation"}),
+            )
+            # "sw5n-wg2p" only:
+            # this dataset has an error: for `demographic_type="overall"`, it has
```

Comment:
Thank you for checking carefully for the quirky issues with each data set!

Reply:
Partially this kind of thing is checked for by the validation function

But there are some other tricky bits that I got to by manually examining the data

It might be worth expanding the cleaning demo, to make this into a turn-the-crank, step-by-step process